File: DSA problems in c++/Merge sort.cpp
#include<iostream>
using namespace std;

void merge(int* arr, int s, int e)
{
    int mid = s + (e - s)/2;

    int len1 = mid - s + 1;
    int len2 = e - mid;

    int* first = new int[len1];
    int* second = new int[len2];

    //copy values
    int mainarrayindex = s;
    for(int i = 0; i < len1; i++)
    {
        first[i] = arr[mainarrayindex++];
    }

    mainarrayindex = mid + 1;
    for(int i = 0; i < len2; i++)
    {
        second[i] = arr[mainarrayindex++];
    }

    //merge 2 two sorted array
    int index1 = 0;
    int index2 = 0;
    mainarrayindex = s;

    while(index1 < len1 && index2 < len2)
    {
        if(first[index1] < second[index2])
            arr[mainarrayindex++] = first[index1++];
        else
            arr[mainarrayindex++] = second[index2++];
    }

    while(index1 < len1)
    {
        arr[mainarrayindex++] = first[index1++];
    }

    while(index2 < len2)
    {
        arr[mainarrayindex++] = second[index2++];
    }
    
    delete []first;
    delete []second;
}

void mergesort(int* arr, int s, int e)
{
    //base case
    if(s >= e)
        return;

    int mid = s + (e - s)/2;
    //left part sort
    mergesort(arr, s, mid);

    //right part sort
    mergesort(arr, mid+1, e);

    //merge the arrays
    merge(arr, s, e);
}

int main()
{
    int arr[5] = {8,3,2,5,3};

    int n = 5;

    mergesort(arr, 0, n-1);

    for(int i = 0; i < n; i++)
        cout << arr[i] << " ";
    
    return 0;
}

File: Program's_Contributed_By_Contributors/Sorting_Algorithms/MergeSort.java
// Java program for Merge Sort
import java.io.*;
import java.util.Scanner;

class MergeSort {

	// Merges two subarrays of arr[].
	// First subarray is arr[l..m]
	// Second subarray is arr[m+1..r]
	void merge(int arr[], int l, int m, int r)
	{
		// Find sizes of two subarrays to be merged
		int n1 = m - l + 1;
		int n2 = r - m;

		// Create temp arrays
		int L[] = new int[n1];
		int R[] = new int[n2];

		// Copy data to temp arrays
		for (int i = 0; i < n1; ++i)
			L[i] = arr[l + i];
		for (int j = 0; j < n2; ++j)
			R[j] = arr[m + 1 + j];

		// Merge the temp arrays

		// Initial indices of first and second subarrays
		int i = 0, j = 0;

		// Initial index of merged subarray array
		int k = l;
		while (i < n1 && j < n2) {
			if (L[i] <= R[j]) {
				arr[k] = L[i];
				i++;
			}
			else {
				arr[k] = R[j];
				j++;
			}
			k++;
		}

		// Copy remaining elements of L[] if any
		while (i < n1) {
			arr[k] = L[i];
			i++;
			k++;
		}

		// Copy remaining elements of R[] if any
		while (j < n2) {
			arr[k] = R[j];
			j++;
			k++;
		}
	}

	// Main function that sorts arr[l..r] using
	// merge()
	void sort(int arr[], int l, int r)
	{
		if (l < r) {

			// Find the middle point
			int m = l + (r - l) / 2;

			// Sort first and second halves
			sort(arr, l, m);
			sort(arr, m + 1, r);

			// Merge the sorted halves
			merge(arr, l, m, r);
		}
	}
	static void printArray(int arr[])
	{
		int n = arr.length;
		for (int i = 0; i < n; ++i)
			System.out.print(arr[i] + " ");
		System.out.println();
	}

	
	public static void main(String args[])
	{
		Scanner sc = new Scanner(System.in);

		System.out.print("Enter the length of Array = ");
		int n = sc.nextInt();
		int arr[] = new int[n];
		System.out.print("\n Enter the values of array = ");

		for(int i =0; i<n;i++){
			System.out.print("\n Enter the value of Array at index "+ i +" = ");
			arr[i] =sc.nextInt();
		}
		

		System.out.println("Given array is");
		printArray(arr);

		MergeSort ob = new MergeSort();
		ob.sort(arr, 0, arr.length - 1);

		System.out.println("\nSorted array is");
		printArray(arr);
	}
}
// This code is contributed
// by Vinayak kushwah
// #Hacktoberfest2023